plugins/uefi-capsule/fu-uefi-device.c

```diff
@@ -754,30 +755,27 @@ fu_uefi_device_class_init (FuUefiDeviceClass *klass)
 }
 
 FuUefiDevice *
-fu_uefi_device_new_from_entry (const gchar *entry_path, GError **error)
+fu_uefi_device_new_from_entry (FuUefiEsrtEntry *entry, GError **error)
 {
 	g_autoptr(FuUefiDevice) self = NULL;
-	g_autofree gchar *fw_class_fn = NULL;
 	g_autofree gchar *id = NULL;
 
-	g_return_val_if_fail (entry_path != NULL, NULL);
+	g_return_val_if_fail (entry != NULL, NULL);
 
 	/* create object */
 	self = g_object_new (FU_TYPE_UEFI_DEVICE, NULL);
 
 	/* assume a uint64_t unless told otherwise by a quirk entry or metadata */
 	fu_device_set_version_format (FU_DEVICE (self), FWUPD_VERSION_FORMAT_NUMBER);
 
-	/* read values from sysfs */
-	fw_class_fn = g_build_filename (entry_path, "fw_class", NULL);
-	if (g_file_get_contents (fw_class_fn, &self->fw_class, NULL, NULL))
-		g_strdelimit (self->fw_class, "\n", '\0');
-	self->capsule_flags = fu_uefi_read_file_as_uint64 (entry_path, "capsule_flags");
-	self->kind = fu_uefi_read_file_as_uint64 (entry_path, "fw_type");
-	self->fw_version = fu_uefi_read_file_as_uint64 (entry_path, "fw_version");
-	self->last_attempt_status = fu_uefi_read_file_as_uint64 (entry_path, "last_attempt_status");
-	self->last_attempt_version = fu_uefi_read_file_as_uint64 (entry_path, "last_attempt_version");
-	self->fw_version_lowest = fu_uefi_read_file_as_uint64 (entry_path, "lowest_supported_fw_version");
+	self->fw_class = fu_uefi_esrt_entry_get_class (entry);
+
+	self->capsule_flags = fu_uefi_esrt_entry_get_capsule_flags (entry);
+	self->kind = fu_uefi_esrt_entry_get_kind (entry);
+	self->fw_version = fu_uefi_esrt_entry_get_version (entry);
```

Comment:
could we just store a ref to the FuUefiEsrtEntry?

Reply:
This is just one of three constructors, storing a ref would affect usage of the data a lot.

Reply:
>  storing a ref would affect usage of the data a lot

How do you mean? I think it would make things a lot simpler.

Reply:
You'd have to check how device was constructed before accessing its data, so every data access would become conditional. I assumed you suggest to store the ref instead of copying its data or maybe I misunderstood?

Reply:
I think fu_uefi_esrt_setup() should read out all the properties on the ESRT node and allocate strings. Then `fu_uefi_esrt_entry_get_class()` should return a `const gchar *` -- i.e. all the failable i/o should be done in _setup() and then the getters become inexpensive accessors into the data.

Reply:
It might make sense to define a parent GObject with all the logic, and then subclass from that for the linux/freebsd versions just with a ->setup() vfunc. If you'd like me to set that up for you with an example, can you push your changes to origin (i.e. I'll make you a committer) and then I can edit your branch too.

Reply:
Thanks, you can push to my branch anyway, because "Allow edits by maintainers" is set for this pull request. Pushed `wip/3mdeb/freebsd-esrt` branch if that's more convenient.

> If you'd like me to set that up for you with an example

That would be helpful. You could also point to a similar case in the sources.

Reply:
Okay, I got carried away and ended up adding a bit more scaffolding as I don't think ESRT will be the only case like this. Alternative proposal in https://github.com/fwupd/fwupd/pull/3289 -- comments welcome.